### ISSUE TITLE
[PatternMatchingInstanceof] False-positives

### ISSUE THREAD

**Issue by boris-petrov** [google/error-prone#4421]
```java
public interface MyMap<K, V> {
  default void foo(Object object) {
    if (object instanceof MyMap) {
      var map = (MyMap<K, V>) object;
    }
  }
}
```
This code gives a `PatternMatchingInstanceof` warning but that's not true when generics are involved.
